### ISSUE TITLE
[RFC] Domain.Sync.notify does it need to be synchronous?

### ISSUE THREAD

**Issue by ctk21** [ocamllabs/ocaml-multicore#386]
In `Domain.Sync.notify` we wait for the notified domain to complete its critical section before it returns:
```ocaml
  val notify : id -> unit
  (** If the domain [d] is within a critical section (i.e. is evaluating
      [critical_section f]), then [notify d] marks this critical section
      as "notified" (causing any call to [wait] to return, see below),
      and waits for the critical section to complete before returning.
      If [d] is not in a critical section, then [notify d] does nothing. *)
```
 I notice in the documentation we have this as a deliberate design:
```
      The only difference from standard monitors is that [notify d]
      waits for any in-progress critical section to complete.
```
I’m not sure why we force notify to be synchronous with the completion of the critical section.

It being synchronous is also a source of bugs, for example this (slightly contrived) example will deadlock:
```ocaml
let _ =
  Domain.Sync.critical_section( fun () ->
    Domain.Sync.notify (Domain.self ()));
  Printf.printf "OK\n"
```
You could imagine other cases where two domains are in critical sections and signalling each other which could deadlock. 

I'm in favour of making `notify` asynchronous (and non-blocking), for a couple of reasons:
- improving the performance when there are large numbers of domains to signal
- less surprise for those who have used pthread `signal`
- reducing scope for deadlocks

However are there usages of the current synchronous implementation that we should be aware of?

**Comment by kayceesrk** [comment]
@stedolan any thoughts on this one?

**Comment by ctk21** [comment]
It looks like we are using the synchronous nature of notify to implement `Domain.join` here:
https://github.com/ocaml-multicore/ocaml-multicore/blob/3d2fc0afb83dfab35c7d87a967cd60629e892bf7/stdlib/domain.ml#L102-L105

Looking in the history, this looks like it was introduced to make a strong reuse guarantee between `join` returning and a subsequent `spawn` with this commit https://github.com/ocaml-multicore/ocaml-multicore/commit/76c8cb90408bd69ae31210c29e5058c3a1928e08

So if `notify` is going to be asynchronous we would need another way to do `Domain.join` if we wanted to support the strong reuse property.

**Comment by stedolan** [comment]
Here's some design notes on Domain.Sync. (Half-remembered, so take with a pinch of salt).

The goal of the Domain.Sync API was to (a) codify a particular mode of use of condition variables and mutexes, and (b) be implementable efficiently enough to support some unusual techniques such as biased locking and RCU. I didn't end up having much time to experiment with biased locking and RCU, so if you were to rip out this support and replace it with something more standard then I don't think anything would really be lost. Still, I'll explain the original design goals here.

## Messages via condition variables

Here's the basic pattern for sending a one-shot message from one pthread to another using condition variables and mutexes:
```c
// receiving thread
pthread_mutex_lock(&lock);
while (!message_flag) {
  pthread_cond_wait(&cond, &lock);
}
pthread_mutex_unlock(&lock);

// sending thread
pthread_mutex_lock(&lock);
message_flag = 1;
pthread_cond_signal(&cond);
pthread_mutex_unlock(&lock);
```
There is an annoying performance issue here, depending on exactly how pthreads is implemented: the sending thread's signal can wake the receiving thread, which immediately sleeps again contending the lock. This causes several more kernel-level context-switches than necessary.

Instead, we can safely move the call to `pthread_cond_signal` below the critical section, which ensures that by the time the receiver has woken the sender has already released the lock. This introduces more nondeterminism: it is now possible for the receiving thread to take the lock between the setting of the message flag and the sending of the signal. This is fine: the signal is ignored, but the message gets through regardless.

By similar reasoning, we can safely move the setting of `message_flag` *above* the critical section, assuming it is a suitable atomic (RA or SC). The sending-side code now looks like:
```c
message_flag = 1;
pthread_mutex_lock(&lock);
pthread_mutex_unlock(&lock);
pthread_cond_signal(&cond);
```
Locking and immediately unlocking the mutex looks like a no-op, but is not. If we remove the lock and unlock, this code no longer works because the following sequence of events can occur:
```c
/* sender */                /* receiver */
                            pthread_mutex_lock(&lock);
                            while (!message_flag) {
message_flag = 1;
pthread_cond_signal(&cond);
                              pthread_cond_wait(&cond, &lock);
                              // deadlocked
```
The lock-unlock sequence in the sender mean that the receiver's check-wait sequence happens either entirely before the lock-unlock (and so the receiver will catch the signal and wake up) or entirely after the lock-unlock sequence (and so the receiver will see the message flag set).

The lock-unlock sequence is what in RCU terms would be called a _quiescent point_: all receiver critical sections must happen entirely before or entirely after this point.

## Domain.Sync

The operations that Domain.Sync makes available (critical_section, wait and notify) are exactly the ones used above, with one lock and condition variable per domain. In principle, they have the following semantics:
```c
critical_section f:
  pthread_mutex_lock(&lock); f (); pthread_mutex_unlock(&lock);
wait:
  pthread_cond_wait(&cond, &lock);
notify D:
  pthread_mutex_lock(&D->lock); pthread_mutex_unlock(&D->lock); pthread_cond_signal(&D->cond);
```
Note the lock-unlock sequence in notify: without it, these operations cannot be used to send messages between domains due to the deadlock bug described above.

The actual implementation optimises for `critical_section`, on the basis that a domain could be constantly polling for a message that's only sent once. This implementation uses a _nonatomic_ variable for the critical section lock, and `notify` is implemented by sending a message to the target domain and waiting for a response. This message is replied to immediately if the target is not in a critical section, but if it is then it waits until the end of the critical section before replying.

## Biased locks and RCU

Biased locks and RCU are two highly efficient synchronisation mechanisms that are not widely used in user-level code in practice, due to the difficulty of integrating them with most systems. Essentially, both of them require a guarantee of regular polling, and this is impossible to make in most languages. (One exception is that the Linux kernel is a very heavy user of RCU. In kernel-mode, the guarantees RCU needs are easier to make). Since we need to ensure regular polling anyway for multicore OCaml, it seems like a good opportunity to try these mechanisms out. Again, I never really got around to this part, so I don't have hard numbers on how it pans out.

### Biased locks

The idea of biased locking is to have locks that are preferentially associated with a single owner thread, and when the owner locks and unlocks them no atomics are required. If they are ever locked from another thread, they get immediately upgraded into a normal lock, removing the biasing ("debiasing"). (There are various schemes for how this should happen, whether they should ever rebias, and so on. I recommend reading at least the intro of [this paper](https://dl.acm.org/doi/pdf/10.1145/2400682.2400686) for more).

The tricky bit is how to do the debiasing. The lock is a single word of state, which the other thread can CAS to mark that it's no longer biased. However, since the owner thread is not using atomics to access the lock, this CAS can be overwritten by a nonatomic access of the owner, leaving two threads in the critical section.

Using Domain.Sync, we can implement biased locks: the owner thread always does the locking inside a Domain.Sync.critical_section. If another thread wants to debias, it performs a CAS and then does a Domain.Sync.interrupt on the owner thread. This ensures that the owner thread passes a quiescent point before the other thread enters its critical section: either the owner thread will have its critical section entirely before the interrupt returns (in which case it will have clobbered the CAS and the other thread retries), or the owner thread will have its critical section entirely after the interrupt returns (in which case nothing is clobbered and the owner thread notices the CAS).

Note that for this to actually be a biased lock, it's essential that we implement Domain.Sync.critical_section using the efficient no-atomics approach described above.

### RCU

RCU (or read-copy-update) is a technique for dealing with shared mutable state that is read much more often than it is written. The standard approach is to use a reader-writer lock, but reader-writer locks involve atomic mutations to the lock state even on the reader path, meaning multiple readers contend on the lock's cache line.

RCU, by contrast, maintains the state as a single atomic word (e.g. pointing to some larger structure), and writers atomically swap it out for a fresh one. The tricky part is determining when all of the readers are done with the old one.

Using Domain.Sync, if all RCU readers do so in a Domain.Sync.critical_section, then the writers can send a `notify` to every other domain after swapping the states. Once the notifications are complete, there can be no readers left using the old state: every reader that began before the notification has completed, and every reader that began afterwards sees the new state.

---

I'm not saying you should necessarily keep this mechanism, or that you should implement biased locking and RCU, but hopefully that answers some of the "wtf is going on here"-type questions! In particular, because of the deadlock issue, simply making notify asynchronous and keeping everything else the same is unsafe.

**Comment by kayceesrk** [comment]
Thanks for the response @stedolan. 

I understand that notify cannot be asynchronous in the sense that the notifying thread needs to wait until the target thread is woken up. But is it necessary that the critical section must be completed? In particular, the waiting thread only acknowledges the pending interrupts at the end of the critical section. I wonder whether acknowledging the interrupts after returning from `caml_plat_wait` will break safety.

**Comment by stedolan** [comment]
I think notify could be asynchronous. Actually, thinking about it some more, I think the complexity arises because `Domain.Sync.notify` combines two different operations: the `lock; unlock` sequence and the `cond_signal` call. It might be worthwhile to separate these into two operations: `lock; unlock` will still block until the mutex is available, but `cond_signal` can be asynchronous.

This would simplify the implementation too: only the `lock;unlock` part needs to implement message-passing. The signal and wait operations can be normal pthreads signal and waits.

**Comment by ctk21** [comment]
TL;DR: I think async is fine for a large class of message passing use cases.

Consider the usage of `critical_section` for receiving a message into a variable with `Domain.Sync.wait` and `Domain.Sync.notify`:

```ocaml
 (* waiting domain *)
 Domain.Sync.critical_section (fun () ->
 	while (!message == None) do
 		Domain.Sync.wait ()
 	done
 )

 (* notifying domain *)
 message := Some v;
 Domain.Sync.notify waiting_domain;

```
[This use case is currently used in domainslib for its channels: https://github.com/ocaml-multicore/domainslib/blob/c819fc0bab791f0258d3cb82a0a159844e6afb69/lib/chan.ml]

Breaking this down into the underlying steps inlining those within `domain.c` (particularly `caml_ml_domain_yield`/`wait` and `caml_ml_domain_interrupt`/`notify`), we get
the following psuedo-code:

```C
  /* waiting domain */
  caml_ml_domain_critical_section(+1);
    while (message != None) {
      pthread_mutex_lock(&waiters_interrupt_lock);
        while (!domain_notified) {
       	  if (handle_incoming_interrupts() == 0)
      	    pthread_cond_wait(&waiters_interrupt_cond);
        }
      pthread_mutex_unlock(&waiters_interrupt_lock);
    }
  caml_ml_domain_critical_section(-1);
  if (critical_section_nesting == 0) {
  	acknowledge_all_pending_interrupts();
  	Assert(domain_notified == 0);
  }

  /* notifying domain */
  message = Some(v);
  pthread_mutex_lock(&waiters_interrupt_lock);
    ... /* enqueue signal interrupt */
    pthread_cond_signal(&waiters_interrupt_cond);
  pthread_mutex_unlock(&waiters_interrupt_lock);
  caml_wait_interrupt_acknowledged(...);

```
NB: the flag `domain_notified` will be set by the service of the signal interrupt by the waiting domain, when it receives the interrupt in `handle_incoming_interrupts` _and_ `critical_section_nesting > 0`.

Currently the notifying domain will only be unblocked from the `caml_wait_interrupt_acknowledged` call once the `acknowledge_all_pending_interrupts` has taken place in the waiting domain.

In the message passing example above, I don't believe that there can be a deadlock if we remove `caml_wait_interrupt_acknowledged`. The reasoning is that:
 - notice that the integrity of the lock/unlock/signal for interrupts ensure that the interrupt itself can never be lost. The waiting domain will always service the interrupt from the notifying domain.
 - `domain_notified` is private to the waiting domain, so has a single reader and writer.
 - within the `caml_ml_domain_critical_section(+1)`/`caml_ml_domain_critical_section(-1)` block in the waiting domain, any servicing of the notify interrupt will result in `domain_notified` being set and checked.

So in the message passing example, we can be asynchronous. However, there are a collection of uses that could do things after the return of the `Domain.Sync.wait` which assume that the notifying domain is not able to touch them or use the return of `Domain.Sync.notify` to know that the waiting domain has reached a certain point of execution (e.g. our implementation of `Domain.join`).

One route to go might be to introduce a flag to notify which denotes waiting or async vs the waiting domain's `critical_section`?

The other, I think is @stedolan's suggestion of breaking notify into 'notify barrier' `lock; unlock` (which presumably includes the interrupt enqueue?) and 'notify signal' `signal`. I'm a bit concerned that having 'notify barrier' and 'notify signal' might make the API more dangerous if people forget to call 'notify signal'. Also wouldn't making the 'wait' a standard pthread wait push the waiting domain into an extra pthread synchronization to wake up; the notify will go to the target pthread but then will need synchronizing with the backup thread to re-enter ocaml code via `caml_leave_blocking_section`. I think I'm probably missing something in the proposal?

**Comment by ctk21** [comment]
I think there is a way to avoid the extra pthread synchronization when the main pthread is re-entering OCaml; so that objection is probably mute: #389

**Comment by stedolan** [comment]
@ctk21 That design makes sense to me (although there's a couple of details about exactly when `domain_notified` gets set that I didn't quite follow). I think the important thing is that conceptually, the target of a `notify` is a critical section, rather than a particular call to `wait`. If `notify` targets `wait` calls, it's extremely tricky to avoid races.

(I am also increasingly suspicious of my earlier choice to allow critical sections to nest. Would it simplify things to make entering a critical section inside a critical section just error? I don't see a use for this ability)

**Comment by ctk21** [comment]
This RFC is now obsolete; `wait`/`notify` have been removed from `Domain` with #561. 
See also #387 for more discussion on why.
